### PR TITLE
When running MSBuild static graph restore display additional log messages when project format is not known to MSBuild

### LINKED ISSUE #10363: StaticGraph Restore should clearly call out when a project is not `KnownToBeMSBuildFormat` When Called via a Solution File
I think is where the actual issue is:

https://github.com/NuGet/NuGet.Client/blob/223189bbc21e8259e6771d363e69d37d9b10e693/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs#L508

In addition there is also another bug in here (if you want I can file another bug report):

@jeffkl, what happens if the project doesn't report itself as `SolutionProjectType.KnownToBeMSBuildFormat`?

We had to have our third party submit this PR https://github.com/dotnet/msbuild/pull/5698 (Thanks @madkat) to get this vendor's project extension in the "blessed list" but not all people who extend the project system are going to be as on top of it to actually file a bug, much less a pull request.

This has wide ranging consequences if you intend to make this the default as proposed by #9803 and more discussion at https://github.com/NuGet/docs.microsoft.com-nuget/issues/1941. Note that this is the SAME class of bug as noted here https://github.com/dotnet/msbuild/issues/5159 for full blown /graph that @cdmihai should be aware of as well.

Right or wrong tons of ISV's have extended the project system in unexpected ways, there needs to be a better way to resolve this, the gains **are super worth it** (@madkat please feel free to share the gains you've seen on our code base and others if you can), and we are **super grateful** that you guys are pushing this forward, but there are growing pains for sure.

_Originally posted by @aolszowka in https://github.com/NuGet/Home/issues/10307#issuecomment-738505272_

@jeffkl Found an old Visual Studio box that doesn't have the pulled PR; compare the following (same project/solution I probably messed up the xxxx in the sanitation between the two so ignore that):

## Visual Studio 2019 MSBuild 16.7.4:
```
R:\xxxxxx\xxxxxx\xxxxxx\xxxxxx\xxxxxx>msbuild.exe /t:restore xxxxxxxxxxxxx.xxxxxxx.sln /p:RestoreUseStaticGraphEvaluation=true
Microsoft (R) Build Engine version 16.7.0+b89cb5fde for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 12/4/2020 12:41:18 PM.
Project "R:\xxxxxx\xxxxxx\xxxxxx\xxxxxx\xxxxxx\xxxxxxxxxxxxx.xxxxxxx.sln" on node 1 (Restore target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Any CPU".
Restore:
  Determining projects to restore...
  Evaluated 0 project(s) in 142ms (0 builds, 0 failures).
Done Building Project "R:\xxxxxx\xxxxxx\xxxxxx\xxxxxx\xxxxxx\xxxxxxxxxxxxx.xxxxxxx.sln" (Restore target(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.01
```

## Visual Studio 2019 MSBuild 16.8.2 (this has the patch applied to allow this third party type to be recongized as a `KnownToBeMSBuildFormat` type:
```
S:\XXXXXX\xx\xxxxx\xxxxxx\xxxxxxxxx\xxxxxxxxxxxxxxx\xxxxxx>msbuild.exe /Target:Restore xxxxxxxxxxxxx.xxxxxxx.sln /p:RestoreUseStaticGraphEvaluation=true
Microsoft (R) Build Engine version 16.8.2+25e4d540b for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 12/4/2020 12:41:28 PM.
Project "S:\XXXXXX\xx\xxxxx\xxxxxx\xxxxxxxxx\xxxxxxxxxxxxxxx\xxxxxx\xxxxxxxxxxxxx.xxxxxxx.sln" on node 1 (Restore target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Any CPU".
Restore:
  Determining projects to restore...
  Evaluated 3 project(s) in 703ms (3 builds, 0 failures).
  Committing restore...
  Assets file has not changed. Skipping assets file writing. Path: S:\XXXXXX\xx\xxxxx\xxxxxx\xxxxxxxxx\xxxxxxxxxxxxxxx\xxxxxx\obj\project.a
  ssets.json
  Restored S:\XXXXXX\xx\xxxxx\xxxxxx\xxxxxxxxx\xxxxxxxxxxxxxxx\xxxxxx\xxxxxxxxxxxxx.xxxxxxx.synproj (in 70 ms).

  NuGet Config files used:
      S:\XXXXXX\xx\xxxxx\NuGet.Config
      C:\Users\aceo\AppData\Roaming\NuGet\NuGet.Config
      C:\xxxxxx Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

  Feeds used:
      S:\XXXXXX\xx\xxxxx\ProductDependencies\_Packages
      https://api.nuget.org/v3/index.json
      C:\xxxxxx Files (x86)\Microsoft SDKs\NuGetPackages\
      https://dotnet.myget.org/F/roslyn/api/v3/index.json
  All projects are up-to-date for restore.
Done Building Project "S:\XXXXXX\xx\xxxxx\xxxxxx\xxxxxxxxx\xxxxxxxxxxxxxxx\xxxxxx\xxxxxxxxxxxxx.xxxxxxx.sln" (Restore target(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.91
```

As I described above, if you don't register yourself as `KnownToBeMSBuildFormat` you get lost when it attempts to perform a Graph based restore.

For our particular usecase with our third party we're OK, **how do other ISV's know to do this?**

_Originally posted by @aolszowka in https://github.com/NuGet/Home/issues/10307#issuecomment-738983763_

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10363

Regression? Last working version:

## Description

MSBuild static graph restore silently ignores projects in solution whose format is not known to MSBuild. Added following log messages which will be displayed at `Normal` verbosity.

- `The solution did not have any projects to restore, ensure that all projects are known to be MSBuild and that the projects exist. ` message will be logged when solution file doesn't have any projects to restore.
- `The solution contains {count} project(s) "{list of projects}" that are not known to MSBuild. Ensure that all projects are known to be MSBuild before running restore on the solution. ` message will be logged when a solution has projects whose format is not known to MSBuild.

For some reason, MSBuild verbosity argument is not passed to NuGet. Hence NuGet always runs static graph restore logic in `Normal` verbosity. It is very much possible that I am missing something. Happy to learn based on the feedback from the team. https://github.com/NuGet/NuGet.Client/blob/5d43a9c072da4cdc4ceee3a75946bf5e2f111a0f/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs#L59

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added

- **Documentation**
  
  - [x] N/A

### CODE REVIEW

**Comment by nkolev92** [comment]
> For some reason, MSBuild verbosity argument is not passed to NuGet. Hence NuGet always runs static graph restore logic in Normal verbosity. It is very much possible that I am missing something. Happy to learn based on the feedback from the team.

If you look at https://github.com/nuget/home/issues?q=is%3Aopen+is%3Aissue+label%3AArea%3ARestoreStaticGraph, you will find https://github.com/NuGet/Home/issues/9026, which calls out this problem.

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:110]: Is this something that should maybe be logged in the BuildTasksUtility instead?

Is the behavior among static and non-static graph equivalent?

File: src/NuGet.Core/NuGet.Build.Tasks/Strings.resx
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks/Strings.resx]: Add  '{0}' and '{1}'. 

I think a lot of the other messages use that approach.

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:558]: I think this should be added at a high verbosity, such as debug. 

Information is too noisy imo.

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:558]: I agree but I am having tough time in adding tests at `Debug/Diagnostic` verbosity because as you know static graph restore defaults to "Normal" verbosity.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:558]: hmm, so we already do the same with standard restore. 
NuGet logs everything, and MSBuild decides whether it will surface it to the user. 
I think static graph can follow the same approach.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:558]: Consider this resolved for now.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:558]: Please add a comment in the linked issue about this.

File: src/NuGet.Core/NuGet.Build.Tasks/Strings.resx
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Build.Tasks/Strings.resx]: Fixed in the latest commit.

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:110]: `BuildTasksUtility` currently doesn't log any message when there are no projects to restore.
https://github.com/NuGet/NuGet.Client/blob/a5d5e3236e86f6c3c09097b723ce2b52852a1752/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs#L234

AFAIK, `dependencyGraphSpec` model doesn't know whether `solution file` or `project file` passed to restore command. `BuildTasksUtility` also doesn't have access to the `entryProjectFilePath` argument passed to static graph restore.

Hence this log message can't be added in `BuildTasksUtility` for non-static graph equivalent restore (for e.g. dotnet restore). I could be wrong but happy to learn more incase if I am missing something.

**Review by nkolev92 — APPROVED** [review]

**Comment by kartheekp-ms** [comment]
@NuGet/nuget-client team, I am waiting for second approval or additional feedback.

**Pull request merged**
